### PR TITLE
Fix panic when reporting shard diagnostics

### LINKED ISSUE #2421: Self-monitoring panic after inserting data
Built from source at 615aabf2fc5df52dc2d011972b53046d240cd10c.

After starting InfluxDB with an empty data directory, running this command:

```
curl -G http://localhost:8086/query --data-urlencode "q=CREATE DATABASE foo" && \
  curl -d '{"database":"foo", "points": [{"name": "cpu","fields": {"value": 100}}]}' -H "Content-Type: application/json" http://localhost:8086/write
``` 

causes the server to panic:

```
[messaging] 2015/04/24 14:17:14 connected to broker: http://:8086/messaging/messages?index=0&streaming=true&topicID=1
[server] 2015/04/24 14:17:14 retention policy 'default' auto-created for database 'foo'
[http] 2015/04/24 14:17:14 ::1 - - [24/Apr/2015:14:17:14 -0700] GET /query?q=CREATE%20DATABASE%20foo HTTP/1.1 200 16 - curl/7.37.1 476a712d-eac7-11e4-800b-000000000000 64.122012ms
[http] 2015/04/24 14:17:14 ::1 - - [24/Apr/2015:14:17:14 -0700] POST /write HTTP/1.1 200 0 - curl/7.37.1 4775dbf6-eac7-11e4-800c-000000000000 191.355976ms
[messaging] 2015/04/24 14:17:14 connected to broker: http://:8086/messaging/messages?index=0&streaming=true&topicID=2
panic: runtime error: index out of range

goroutine 259 [running]:
github.com/influxdb/influxdb.(*Server).convertRowToPoints(0xc20805ab00, 0x5e4b70, 0xb, 0xc2081233e0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/Users/mr/go/src/github.com/influxdb/influxdb/server.go:4018 +0x7b7
github.com/influxdb/influxdb.func·011()
	/Users/mr/go/src/github.com/influxdb/influxdb/server.go:399 +0x5c0
created by github.com/influxdb/influxdb.(*Server).StartSelfMonitoring
	/Users/mr/go/src/github.com/influxdb/influxdb/server.go:412 +0x241
```

### PR DESCRIPTION
**Pull request by marcosnils**
Shard diagnostics wasn't been populated correctly and it made influx panic when reporting shards.
I also changed the TestSingleServerDiags test to reproduce the error
Fixes #2421

### CODE REVIEW

File: server.go [hunk 1]
```diff
@@ -3345,9 +3346,14 @@ func (s *Server) DiagnosticsAsRows() []*influxql.Row {
 
 		// Shard may not be local to this node.
 		if sh.store != nil {
-			shardsRow.Columns = append(shardsRow.Columns, "path")
-			shardsRow.Values[0] = append(shardsRow.Values[0], sh.store.Path())
+			if !hasPath {
```
Comment (otoolep) [inline server.go]: I don't think this is necessary. If a `sh.store` is non-nil, then it must have a path.

File: server.go [hunk 2]
```diff
@@ -3345,9 +3346,14 @@ func (s *Server) DiagnosticsAsRows() []*influxql.Row {
...
+				shardsRow.Columns = append(shardsRow.Columns, "path")
+				hasPath = true
+			}
+			index := len(shardsRow.Values) - 1 // Current shardsRow.Values index
+			shardsRow.Values[index] = append(shardsRow.Values[index], sh.store.Path())
```
Comment (otoolep) [inline server.go]: I'm not sure this is right, but I might be missing something. `shardsRow.Values[0]` is a slice -- the code wants to append to this slice, not add another distinct entry to `shardsRow.Values`.

Can you show some raw data to explain what you are trying to fix?

File: cmd/influxd/server_integration_test.go [hunk 3]
```diff
@@ -1462,9 +1462,13 @@ func TestSingleServerDiags(t *testing.T) {
 	config.Monitoring.Enabled = true
 	config.Monitoring.WriteInterval = main.Duration(100 * time.Millisecond)
 	nodes := createCombinedNodeCluster(t, testName, dir, 1, config)
+	createDatabase(t, testName, nodes, "mydb")
```
Comment (otoolep) [inline cmd/influxd/server_integration_test.go]: Why are these changes here?

File: server.go (hunk 1 above)
Comment (marcosnils) [inline server.go]: @otoolep but you iterate through each shard. The hasPath variable controls if if the path column has been added before or we need to add it as it's the first time it appears

File: cmd/influxd/server_integration_test.go (hunk 3 above)
Comment (otoolep) [inline cmd/influxd/server_integration_test.go]: Oh, OK, to repro the error.

Well, let me run your code, and I'll see if I can get answers to my questions below.

Comment (marcosnils) [inline cmd/influxd/server_integration_test.go]: because we need to to actually generate some shard data for the diag module to process it. Before this change the server diagnostic was just processing some basic information but not shard data

**Comment by otoolep** [comment]
You can ignore the test failure for now, it's a known issue we need to look into.

**Comment by otoolep** [comment]
OK, thanks @marcosnils -- I'll take a closer look at this diff shortly.

File: server.go (hunk 2 above)
Comment (marcosnils) [inline server.go]: Sure, before my change the server was doing something like this: 

```shards_diag
[2015-04-25 01:15:03.956810841 +0000 UTC 1 1 550 /home/marcos/.influxdb/data/shards/1 /home/marcos/.influxdb/data/shards/2] map[id:1 dataNodes:2 index:3 path:5]
[2015-04-25 01:15:03.956810841 +0000 UTC 2 1 221] map[index:3 path:5 id:1 dataNodes:2]
```
After my change is doing this


```shards_diag
[2015-04-25 01:15:03.956810841 +0000 UTC 1 1 550 /home/marcos/.influxdb/data/shards/1] map[id:1 dataNodes:2 index:3 path:4]
[2015-04-25 01:15:03.956810841 +0000 UTC 2 1 221 /home/marcos/.influxdb/data/shards/2] map[index:3 path:4 id:1 dataNodes:2]
```
My understanding by looking at the code is that the `shard_diag` row should contain one row per shard, and previously it was appending all the shards in the first row

**Comment by otoolep** [comment]
OK, thanks very much @marcosnils -- I see where you are going. In fact this code has caused multiple issues, so I've actually simplified it at https://github.com/influxdb/influxdb/pull/2431.

Your code showed me what was wrong, but I'd like to solve it slightly differently. Please take a look at PR 2431.
